feat(.NET): Consolidating APIs (#17266)
newrelic/docs-website: src/content/docs/apm/agents/net-agent/net-agent-api/net-agent-api.mdx

---
title: .NET agent API
tags:
  - Agents
  - NET agent
  - API guides

> metaDescription: 'A goal-focused guide to the New Relic .NET agent API.'

Reviewer: This meta-description feels outdated for the current scope of the doc.

  - /docs/agents/net-agent/net-agent-api/seterrorgroupcallback-net-agent
  - /docs/agents/net-agent/net-agent-api/set-error-group-callback
  - /docs/apm/agents/net-agent/net-agent-api/seterrorgroupcallback-net-agent-api
  - /docs/agents/net-agent/net-agent-api/setllmtokencountingcallback-net-agent-api
  - /docs/agents/net-agent/net-agent-api/setllmtokencountingcallback-net-agent
  - /docs/agents/net-agent/net-agent-api/set-llm-token-counting-callback
  - /docs/apm/agents/net-agent/net-agent-api/setllmtokencountingcallback-net-agent-api
  - /docs/agents/net-agent/net-agent-api/settransactionname-net-agent-api
  - /docs/agents/net-agent/net-agent-api/settransactionname-net-agent
  - /docs/agents/net-agent/net-agent-api/set-transaction-name
  - /docs/apm/agents/net-agent/net-agent-api/settransactionname-net-agent-api
  - /docs/agents/net-agent/net-agent-api/start-agent
  - /docs/agents/net-agent/net-agent-api/startagent-net-agent
  - /docs/apm/agents/net-agent/net-agent-api/start-agent
  - /docs/agents/net-agent/net-agent-api/tracemetadata-net-agent-api-0
  - /docs/apm/agents/net-agent/net-agent-api/tracemetadata-net-agent-api-0

> freshnessValidatedDate: never

Reviewer: This should be updated

---

> New Relic's .NET agent includes an API that allows you to extend the agent's standard functionality. For example, you can use the .NET agent API for:

Reviewer: ```suggestion
New Relic's .NET agent includes an API that allows you to extend the agent's standard functionality. For example, you can use the .NET agent API to:
```

> * Customizing your app name

Reviewer: Unnecessary -ing verbs

* Creating custom transaction parameters
* Reporting custom errors and metrics

You can also customize some of the .NET agent's default behavior by adjusting [configuration settings](/docs/agents/net-agent/configuration/net-agent-configuration) or using [custom instrumentation](/docs/agents/net-agent/custom-instrumentation/introduction-net-custom-instrumentation).

## Requirements

To use the .NET agent API, make sure you have the [latest .NET agent release](/docs/release-notes/agent-release-notes/net-release-notes). Then, add a reference to the agent in your project using one of the two options below:

   * Add a reference to `NewRelic.Api.Agent.dll` to your project.

>     OR

Reviewer: I think this needs an extra line break or something to show up on its own line, in preview it shows up on same line

![Uploading image.png…]()

## List of API calls

   The following list contains the different calls you can make with the API, including syntax, requirements, functionality, and examples:

<CollapserGroup>
    <Collapser
        id="DisableBrowserMonitoring"
        title="DisableBrowserMonitoring"
    >
    ### Syntax

        ```cs
        NewRelic.Api.Agent.NewRelic.DisableBrowserMonitoring([boolean $override])
        ```

        Disable automatic injection of browser monitoring snippet on specific pages.

    ### Requirements

    * Compatible with all agent versions.
    * Must be called inside a [transaction](/docs/glossary/glossary/#transaction).

    ### Description

    Add this call to disable the <DoNotTranslate>**automatic**</DoNotTranslate> injection of [<InlinePopover type="browser" />](/docs/browser/new-relic-browser/getting-started/new-relic-browser) scripts on specific pages. You can also add an optional override to disable <DoNotTranslate>**both**</DoNotTranslate> manual and automatic injection. In either case, put this API call as close as possible to the top of the view in which you want browser disabled.

    <Callout variant="tip">

>     Compare [`GetBrowserTimingHeader()`](/docs/agents/net-agent/net-agent-api/getbrowsertimingheader-net-agent), which <DoNotTranslate>**adds**</DoNotTranslate> the browser script to the page.

Reviewer: This was formerly a link to another doc, but now should presumably be an anchor link within the same doc. Would be good to check for other examples where links like this need updating.

>     Compare [`GetBrowserTimingHeader()`](/docs/agents/net-agent/net-agent-api/getbrowsertimingheader-net-agent), which <DoNotTranslate>**adds**</DoNotTranslate> the browser script to the page.

Reviewer: I don't think `**adds**` should have the DNT tag

    </Callout>

    ### Parameters

    <table>
    <thead>
        <tr>
        <th width="25%">
            Parameter
        </th>

        <th>
            Description
        </th>
        </tr>
    </thead>

    <tbody>
        <tr>
        <td>
            `$override`

            _boolean_
        </td>

        <td>
            Optional. When `true`, disables all injection of browser scripts. This flag affects both manual and automatic injection. This also overrides the [`GetBrowserTimingHeader()`](/docs/agents/net-agent/net-agent-api/getbrowsertimingheader-net-agent) call.
        </td>
        </tr>
    </tbody>
    </table>

    ### Examples

>     #### Disable automatic injection [#automatic-only]

Reviewer: These IDs dont work in the preview build, I think b/c h4s arent really supported/preferred 
<img width="437" alt="image" src="https://github.com/newrelic/docs-website/assets/55203603/204d7cc3-d16d-4079-adfa-0de5eef76fdd">

    ```

    Get access to the agent via the `IAgent` interface.

    ### Requirements

    * Agent version 8.9 or higher.
    * Compatible with all app types.

    ### Description

    Get access to agent API methods via the [`IAgent`](/docs/agents/net-agent/net-agent-api/iagent) interface.

    ### Return values

    An implementation of [IAgent](/docs/agents/net-agent/net-agent-api/iagent-net-agent-api) providing access to the [IAgent API](/docs/agents/net-agent/net-agent-api/iagent-net-agent-api).

    ### Examples

    ```cs
    IAgent agent = NewRelic.Api.Agent.NewRelic.GetAgent();
    ```

    </Collapser>
    <Collapser
        id="GetBrowserTimingHeader"
        title="GetBrowserTimingHeader"
    >
    ### Syntax

    ```cs
    NewRelic.Api.Agent.NewRelic.GetBrowserTimingHeader();
    NewRelic.Api.Agent.NewRelic.GetBrowserTimingHeader(string nonce);
    ```

    Generate a browser monitoring HTML snippet to instrument end-user browsers.

    ### Requirements

    * Compatible with all agent versions.
    * Must be called inside a [transaction](/docs/accounts-partnerships/education/getting-started-new-relic/glossary#transaction).

    ### Description

>     Returns an HTML snippet used to enable [<InlinePopover type="browser" />](/docs/browser/new-relic-browser/getting-started/new-relic-browser). The snippet instructs the browser to fetch a small JavaScript file and start the page timer. You can then insert the returned snippet into the header of your HTML webpages. For more information, see [Adding apps to Browser monitoring](/docs/browser/new-relic-browser/installation-configuration/adding-apps-new-relic-browser).

Reviewer: ```suggestion
    Returns an HTML snippet used to enable <InlinePopover type="browser" />. The snippet instructs the browser to fetch a small JavaScript file and start the page timer. You can then insert the returned snippet into the header of your HTML webpages. For more information, see [Adding apps to browser monitoring](/docs/browser/new-relic-browser/installation-configuration/adding-apps-new-relic-browser).
```

Reviewer: Two small errors here

    <Callout variant="tip">

>     Compare [`DisableBrowserMonitoring()`](/docs/agents/net-agent/net-agent-api/disablebrowsermonitoring-net-agent), which <DoNotTranslate>**disables**</DoNotTranslate> the browser script on a page.

Reviewer: Same link issue

    </Callout>

    ### Parameters

    <table>
    <thead>
        <tr>
        <th width="25%">
            Parameter
        </th>

        <th>
            Description
        </th>
        </tr>
    </thead>

    <tbody>
        <tr>
        <td>
            `nonce`

            _string_
        </td>

        <td>

>             The per-request, cryptographic nonce used by Content-Security-Policy policies.

Reviewer: ```suggestion
            The per-request, cryptographic nonce used by Content Security Policy policies.
```

        </td>
        </tr>
    </tbody>
    </table>

    <Callout variant="tip">
    This API call requires updates to security allow lists. For more information about Content Security Policy (CSP) considerations, visit the [browser monitoring compatibility and requirements](/docs/browser/new-relic-browser/getting-started/compatibility-requirements-browser-monitoring) page.
    </Callout>

    ### Return values

    An HTML string to be embedded in a page header.

    ### Examples

>     #### With ASPX [#aspx]

Reviewer: Same h4 header ID issue

    </Callout>

    <Callout variant="important">
    This API can not be placed in a `<HeadContent>` element of a `.razor` page.  Instead, it should be called from `_Layout.cshtml` or an equivalent layout file.
    </Callout>

    ```cshtml
    <!DOCTYPE html>
    <html lang="en">
    <head>
        @Html.Raw(NewRelic.Api.Agent.NewRelic.GetBrowserTimingHeader())
        ...
    </head>
    <body>
    ...
    ```

    ```cshtml
    <!DOCTYPE html>
    <html lang="en">
    <head>
        @Html.Raw(NewRelic.Api.Agent.NewRelic.GetBrowserTimingHeader("YOUR_NONCE_VALUE"))
        ...
    </head>
    <body>
    ...
    ```

    </Collapser>
    <Collapser
        id="GetLinkingMetadata"
        title="GetLinkingMetadata"
    >
    ### Syntax

    ```cs
    NewRelic.Api.Agent.NewRelic.GetLinkingMetadata();
    ```

    Returns key/value pairs which can be used to link traces or entities.

    ### Requirements

    * Agent version 8.19 or higher.
    * Compatible with all app types.

    ### Description

>     The Dictionary of key/value pairs returned includes items used to link traces and entities in the APM product. It will only contain items with meaningful values. For instance, if distributed tracing is disabled, `trace.id` will not be included.

Reviewer: Why is `Dictionary` capped?

        </td>
        </tr>

        <tr>
        <td>
            [`AcceptDistributedTracePayload` (obsolete)](#acceptdistributedtracepayload)
        </td>

        <td>
            Accepts an incoming distributed trace payload from another service.
        </td>
        </tr>

        <tr>
        <td>
            [`CreateDistributedTracePayload` (obsolete)](#createdistributedtracepayload)
        </td>

        <td>
            Creates a distributed trace payload for inclusion in an outgoing request.
        </td>
        </tr>

        <tr>
        <td>
            [`AddCustomAttribute`](#addcustomattribute)
        </td>

        <td>
            Add contextual information from your application to the current transaction in form of attributes.
        </td>
        </tr>

        <tr>
        <td>
            [`CurrentSpan`](#currentspan)
        </td>

        <td>
            Provides access to the currently executing [span](/docs/agents/net-agent/net-agent-api/ispan), which provides access to span-specific methods in the New Relic API.
        </td>
        </tr>

        <tr>
        <td>
            [`SetUserId`](#setuserid)
        </td>

        <td>

>             Associates a User Id to the current transaction.

Reviewer: Why is `User Id` capped liek this?

        </td>
        </tr>
    </tbody>
    </table>
    <CollapserGroup>
    <Collapser
        id="AcceptDistributedTraceHeaders"
        title="AcceptDistributedTraceHeaders"
    >

>         `ITransaction.AcceptDistributedTraceHeaders` is used to instrument the called service for inclusion in a distributed trace. It links the spans in a trace by accepting a payload generated by [`InsertDistributedTraceHeaders`](/docs/agents/nodejs-agent/api-guides/nodejs-agent-api#transaction-handle-insertDistributedTraceHeaders) or generated by some other W3C Trace Context compliant tracer. This method accepts the headers of an incoming request, looks for W3C Trace Context headers, and if not found, falls back to New Relic distributed trace headers. This method replaces the deprecated [`AcceptDistributedTracePayload`](/docs/agents/net-agent/net-agent-api/itransaction#acceptdistributedtracepayload) method, which only handles New Relic distributed trace payloads.

Reviewer: For other methods, the syntax appears first and then description. Here description precedes syntax, and there's no short summaruy, only the long description

                Required. Describes the transport of the incoming payload (for example `TransportType.HTTP`).
            </td>
            </tr>
        </tbody>
        </table>

        ### Usage considerations

        * [Distributed tracing must be enabled](/docs/agents/net-agent/configuration/net-agent-configuration#distributed_tracing).
        * `AcceptDistributedTraceHeaders` will be ignored if `InsertDistributedTraceHeaders` or `AcceptDistributedTraceHeaders` has already been called for this transaction.

        ### Example

        ```cs
        HttpContext httpContext = HttpContext.Current;
        IAgent agent = NewRelic.Api.Agent.NewRelic.GetAgent();
        ITransaction currentTransaction = agent.CurrentTransaction;
        currentTransaction.AcceptDistributedTraceHeaders(httpContext, Getter, TransportType.HTTP);
        IEnumerable<string> Getter(HttpContext carrier, string key) 
        { 
            string value = carrier.Request.Headers[key];   
            return value == null ? null : new string[] { value }; 
        }
        ```
    </Collapser>
    <Collapser
        id="InsertDistributedTraceHeaders"
        title="InsertDistributedTraceHeaders"
    >

>         `ITransaction.InsertDistributedTraceHeaders` is used to implement distributed tracing. It modifies the carrier object that is passed in by adding W3C Trace Context headers and New Relic Distributed Trace headers. The New Relic headers can be disabled with `<distributedTracing excludeNewrelicHeader="true" />` in the config. This method replaces the deprecated [`CreateDistributedTracePayload`](/docs/agents/net-agent/net-agent-api/itransaction#createdistributedtracepayload) method, which only creates New Relic Distributed Trace payloads.

Reviewer: Same description issue as other DT method

>         `ITransaction.InsertDistributedTraceHeaders` is used to implement distributed tracing. It modifies the carrier object that is passed in by adding W3C Trace Context headers and New Relic Distributed Trace headers. The New Relic headers can be disabled with `<distributedTracing excludeNewrelicHeader="true" />` in the config. This method replaces the deprecated [`CreateDistributedTracePayload`](/docs/agents/net-agent/net-agent-api/itransaction#createdistributedtracepayload) method, which only creates New Relic Distributed Trace payloads.

Reviewer: ```suggestion
        `ITransaction.InsertDistributedTraceHeaders` is used to implement distributed tracing. It modifies the carrier object that is passed in by adding W3C Trace Context headers and New Relic Distributed Trace headers. The New Relic headers can be disabled with `<distributedTracing excludeNewrelicHeader="true" />` in the config. This method replaces the deprecated [`CreateDistributedTracePayload`](/docs/agents/net-agent/net-agent-api/itransaction#createdistributedtracepayload) method, which only creates New Relic distributed tracing payloads.
```

        ### Syntax

        ```cs
        void InsertDistributedTraceHeaders(carrier, setter)
        ```

        ### Parameters

        <table>
        <thead>
            <tr>
            <th>
                Name
            </th>

            <th>
                Description
            </th>
            </tr>
        </thead>

        <tbody>
            <tr>
            <td>
                `carrier`

                _&lt;T>_
            </td>

            <td>

>                 Required. Container where Trace Context headers are inserted..

Reviewer: ```suggestion
                Required. Container where trace context headers are inserted.
```

            </td>
            </tr>

            <tr>
            <td>
                `setter`

                _Action&lt;T, string, string>_
            </td>

            <td>

>                 Required. Caller-defined Action to insert header data into the carrier.

Reviewer: ```suggestion
                Required. Caller-defined action to insert header data into the carrier.
```

    This method requires .NET agent version and .NET agent API [version 8.24.244.0](/docs/release-notes/agent-release-notes/net-release-notes/net-agent-8242440) or higher. It replaced the deprecated [`AddCustomParameter`](/docs/agents/net-agent/net-agent-api/add-custom-parameter).

    ### Syntax

    ```cs
    ITransaction AddCustomAttribute(string key, object value)
    ```

    ### Parameters

    <table>
    <thead>
        <tr>
        <th>
            Parameter
        </th>

        <th>
            Description
        </th>
        </tr>
    </thead>

    <tbody>
        <tr>
        <td>
            `key`

            _string_
        </td>

        <td>
            Identifies the information being reported. Also known as the name.

            * Empty keys are not supported.
            * Keys are limited to 255-bytes. Attributes with keys larger than 255-bytes will be ignored.
        </td>
        </tr>

        <tr>
        <td>
            `value`

            _object_
        </td>

        <td>
            The value being reported.
            
            <DoNotTranslate>**Note**</DoNotTranslate>: `null` values will not be recorded.
            

>             <table>

Reviewer: This table-in-a-table looks pretty nightmarish in builds, we should find another way to format it
<img width="686" alt="image" src="https://github.com/newrelic/docs-website/assets/55203603/738542eb-67d1-4537-b770-111db290c8b9">

                <td>
                    True or false.
                </td>
                </tr>

                <tr>
                <td>
                    `DateTime`
                </td>

                <td>
                    A string representation following the ISO-8601 format, including time zone information:

                    Example: `2020-02-13T11:31:19.5767650-08:00`
                </td>
                </tr>

                <tr>
                <td>
                    `TimeSpan`
                </td>

                <td>
                    A decimal-based number representing number of seconds.
                </td>
                </tr>

                <tr>
                <td>
                    everything else
                </td>

                <td>
                    The `ToString()` method will be applied. Custom types must have an implementation of `Object.ToString()` or they will throw an exception.
                </td>
                </tr>
            </tbody>
            </table>

        </td>
        </tr>
    </tbody>
    </table>

    ### Returns

    A reference to the current transaction.

    ### Usage considerations

>     For details about supported data types, see the [Custom Attributes Guide](/docs/agents/net-agent/attributes/custom-attributes).

Reviewer: ```suggestion
    For details about supported data types, see [Custom attributes](/docs/agents/net-agent/attributes/custom-attributes).
```

    ### Example

    ```cs
    IAgent agent = NewRelic.Api.Agent.NewRelic.GetAgent();
    ITransaction transaction = agent.CurrentTransaction;
    transaction.AddCustomAttribute("customerName","Bob Smith")
        .AddCustomAttribute("currentAge",31)
        .AddCustomAttribute("birthday", new DateTime(2000, 02, 14))
        .AddCustomAttribute("waitTime", TimeSpan.FromMilliseconds(93842));
    ```
    </Collapser>
    <Collapser
        id="CurrentSpan"
        title="CurrentSpan"
    >
    
    Provides access to the currently executing [span](/docs/apm/agents/net-agent/net-agent-api/ispan/), making [span-specific methods](/docs/apm/agents/net-agent/net-agent-api/ispan) available within the New Relic API.

    ### Example

    ```cs
    IAgent agent = NewRelic.Api.Agent.NewRelic.GetAgent(); 
    ITransaction transaction = agent.CurrentTransaction; 
    ISpan currentSpan = transaction.CurrentSpan;
    ```

    </Collapser>
    <Collapser
        id="SetUserId"
        title="SetUserId"
    >

>     Associates a User Id with the current transaction.

Reviewer: why is `User Id` capped like this?

    ```cs
    Public interfac ISpan
    ```

    Provides access to span-specific methods in the New Relic API.

    ### Description

    Provides access to span-specific methods in the New Relic .NET agent API. To obtain a reference to `ISpan`, use:

    * The `CurrentSpan` property on [`IAgent`](/docs/agents/net-agent/net-agent-api/iagent) (Recommended).
    * The `CurrentSpan` property on [`ITransaction`](/docs/agents/net-agent/net-agent-api/itransaction).

    This section contains descriptions and parameters of `ISpan` methods:

    <table>
    <thead>
        <tr>
        <th>
            Name
        </th>

        <th>
            Description
        </th>
        </tr>
    </thead>

    <tbody>
        <tr>
        <td>
            [`AddCustomAttribute`](#addcustomattribute)
        </td>

        <td>
            Add contextual information from your application to the current span in form of attributes.
        </td>
        </tr>
        <tr>
        <td>
            [`SetName`](#setname)
        </td>

        <td>
            Changes the name of the current span/segment/metrics that will be reported to New Relic.
        </td>
        </tr>
    </tbody>
    </table>

>     ### AddCustomAttribute [#addcustomattribute]

Reviewer: This seems to be missing a collapser to make it its own entry
![Uploading image.png…]()

    This method requires .NET agent version and .NET agent API [version 8.25](/docs/release-notes/agent-release-notes/net-release-notes/net-agent-8242440) or higher.

    ### Syntax

    ```cs
    ISpan AddCustomAttribute(string key, object value)
    ```

    ### Parameters

    <table>
    <thead>
        <tr>
        <th>
            Parameter
        </th>

        <th>
            Description
        </th>
        </tr>
    </thead>

    <tbody>
        <tr>
        <td>
            `key`

            _string_
        </td>

        <td>
            Identifies the information being reported. Also known as the name.

            * Empty keys are not supported.
            * Keys are limited to 255-bytes. Attributes with keys larger than 255-bytes will be ignored.
        </td>
        </tr>

        <tr>
        <td>
            `value`

            _object_
        </td>

        <td>
            The value being reported.
            
            <DoNotTranslate>**Note**</DoNotTranslate>: `null` values will not be recorded.
            

>             <table>

Reviewer: Same table-in-a-table issue as the other attribute method

                <td>
                    A decimal-based number representing number of seconds.
                </td>
                </tr>

                <tr>
                <td>
                    everything else
                </td>

                <td>
                    The `ToString()` method will be applied. Custom types must have an implementation of `Object.ToString()` or they will throw an exception.
                </td>
                </tr>
            </tbody>
            </table>
        </td>
        </tr>
    </tbody>
    </table>

    ### Returns

    A reference to the current span.

    ### Usage considerations

    For details about supported data types, see the [Custom Attributes guide](/docs/agents/net-agent/attributes/custom-attributes).

    ### Examples

    ```cs
    IAgent agent = NewRelic.Api.Agent.NewRelic.GetAgent(); 
    ISpan currentSpan = agent.CurrentSpan; 

    currentSpan
        .AddCustomAttribute("customerName","Bob Smith")
        .AddCustomAttribute("currentAge",31)
        .AddCustomAttribute("birthday", new DateTime(2000, 02, 14))
        .AddCustomAttribute("waitTime", TimeSpan.FromMilliseconds(93842));
    ```

>     ### SetName [#setname]

Reviewer: Missing collapser tags

    </Callout>

    ### Parameters

    <table>
    <thead>
        <tr>
        <th width="25%">
            Parameter
        </th>

        <th>
            Description
        </th>
        </tr>
    </thead>

    <tbody>
        <tr>
        <td>
            `$metric_name`

            _string_
        </td>

        <td>
            Required. The name of the metric to increment.
        </td>
        </tr>
    </tbody>
    </table>

    ### Examples

    ```cs
    NewRelic.Api.Agent.NewRelic.IncrementCounter("Custom/ExampleMetric");
    ```
    </Collapser>

    <Collapser
        id="NoticeError"
        title="NoticeError"
    >
    ### Overloads [#overloads]

    Notice an error and report to New Relic, along with optional custom attributes.

    ```cs
    NewRelic.Api.Agent.NewRelic.NoticeError(Exception $exception);
    NewRelic.Api.Agent.NewRelic.NoticeError(Exception $exception, IDictionary<TKey, TValue> $attributes);
    NewRelic.Api.Agent.NewRelic.NoticeError(string $error_message, IDictionary<TKey, TValue> $attributes);
    NewRelic.Api.Agent.NewRelic.NoticeError(string $error_message, IDictionary<TKey, TValue> $attributes, bool $is_expected);
    ```

>     ### Requirements [#requirements]

Reviewer: ID is bleeding thru in rendered doc
![Uploading image.png…]()

    If this method is invoked within a [transaction](/docs/glossary/glossary/#transaction), the agent reports the exception within the parent transaction. If it is invoked outside of a transaction, the agent creates an [error trace](/docs/errors-inbox/errors-inbox) and categorizes the error in the New Relic UI as a `NewRelic.Api.Agent.NoticeError` API call. If invoked outside of a transaction, the `NoticeError()` call will not contribute to the error rate of an application.

    The agent adds the attributes only to the traced error; it does not send them to New Relic. For more information, see [`AddCustomAttribute()`](/docs/apm/agents/net-agent/net-agent-api/itransaction/#addcustomattribute).

    Errors reported with this API are still sent to New Relic when they are reported within a transaction that results in an HTTP status code, such as a `404`, that is configured to be ignored by agent configuration. For more information, see our documentation about [managing errors in APM](/docs/apm/agents/manage-apm-agents/agent-data/manage-errors-apm-collect-ignore-or-mark-expected).

    Review the sections below to see examples of how to use this call.

>     ### NoticeError(Exception) [#exception-overload]

Reviewer: should this be its own collapser?

    ```cs
    NewRelic.Api.Agent.NewRelic.NoticeError(Exception $exception)
    ```
    <table>
    <thead>
        <tr>
        <th width="25%">
            Parameter
        </th>

        <th>
            Description
        </th>
        </tr>
    </thead>

    <tbody>
        <tr>
        <td>
            `$exception`

            _Exception_
        </td>

        <td>
            Required. The `Exception` you want to instrument. Only the first 10,000 characters from the stack trace are retained.
        </td>
        </tr>
    </tbody>
    </table>

>     ### NoticeError(Exception, IDictionary) [#exception-idictionary-overload]

Reviewer: should this be its own collapser?

    ```cs
    NewRelic.Api.Agent.NewRelic.NoticeError(Exception $exception, IDictionary<TKey, TValue> $attributes)
    ```

    <table>
    <thead>
        <tr>
        <th width="25%">
            Parameter
        </th>

        <th>
            Description
        </th>
        </tr>
    </thead>

    <tbody>
        <tr>
        <td>
            `$exception`

            _Exception_
        </td>

        <td>
            Required. The `Exception` you want to instrument. Only the first 10,000 characters from the stack trace are retained.
        </td>
        </tr>

        <tr>
        <td>
            `$attributes`

            _IDictionary&lt;TKey, TValue&gt;_
        </td>

        <td>
            Specify key/value pairs of attributes to annotate the error message. The `TKey` must be a string, the `TValue` can be a string or object.
        </td>
        </tr>
    </tbody>
    </table>

>     ### NoticeError(String, IDictionary) [#string-idictionary-overload]

Reviewer: should this be its own collapser?

    ```

    <table>
    <thead>
        <tr>
        <th width="25%">
            Parameter
        </th>

        <th>
            Description
        </th>
        </tr>
    </thead>

    <tbody>
        <tr>
        <td>
            `$error_message`

            _string_
        </td>

        <td>
            Required. Specify a string to report to New Relic as though it's an exception. This method creates both [error events](/docs/apm/apm-ui-pages/error-analytics/manage-error-data#events) and [error traces](/docs/apm/apm-ui-pages/error-analytics/manage-error-data#trace-details).  Only the first 1023 characters are retained in error events, while error traces retain the full message.
        </td>
        </tr>

        <tr>
        <td>
            `$attributes`

            _IDictionary&lt;TKey, TValue&gt;_
        </td>

        <td>
            Required (can be null). Specify key/value pairs of attributes to annotate the error message. The `TKey` must be a string, the `TValue` can be a string or object, to send no attributes, pass `null`.  
        </td>
        </tr>
    </tbody>
    </table>

>     ### NoticeError(String, IDictionary, bool) [#string-idictionary-bool-overload]

Reviewer: should this be its own collapser?

    >
    ### Syntax

    ```cs
    NewRelic.Api.Agent.NewRelic.RecordCustomEvent(string eventType, IEnumerable<string, object> attributeValues)
    ```

    Records a custom event with the given name and attributes.

    ### Requirements

    * Agent version 4.6.29.0 or higher.
    * Compatible with all app types.

    ### Description

    Records a [custom event](/docs/data-analysis/metrics/analyze-your-metrics/data-collection-metric-timeslice-event-data#event-data) with the given name and attributes, which you can query in the [query builder](/docs/query-your-data/explore-query-data/query-builder/introduction-query-builder). To verify if an event is being recorded correctly, look for the data in [dashboards](/docs/query-your-data/explore-query-data/dashboards/introduction-new-relic-one-dashboards).

    For related API calls, see the [.NET agent API guide](/docs/agents/net-agent/api-guides/guide-using-net-agent-api#custom-data).

    <Callout variant="important">
    * Sending a lot of events can increase the memory overhead of the agent.
    * Additionally, posts greater than 1MB (10^6 bytes) in size will not be recorded regardless of the maximum number of events.

>     * Custom Events are limited to 64-attributes.

Reviewer: ```suggestion
    * Custom events are limited to 64 attributes.
```

Caps and hyphen here are weird

        </th>
        </tr>
    </thead>

    <tbody>
        <tr>
        <td>
            `$metric_name`

            _string_
        </td>

        <td>
            Required. The name of the response time metric to record. Only the first 255 characters are retained.
        </td>
        </tr>

        <tr>
        <td>
            `$metric_value`

            _Int64_
        </td>

        <td>
            Required. The response time to record in milliseconds.
        </td>
        </tr>
    </tbody>
    </table>

    ### Examples

    #### Record response time of a sleeping process [#record-stopwatch]

    ```cs
    Stopwatch stopWatch = Stopwatch.StartNew();
    System.Threading.Thread.Sleep(5000);
    stopWatch.Stop();
    NewRelic.Api.Agent.NewRelic.RecordResponseTimeMetric("Custom/DEMO_Record_Response_Time_Metric", stopWatch.ElapsedMilliseconds);
    ```

    </Collapser>
    <Collapser
        id="SetApplicationName"
        title="SetApplicationName"
    >
    ### Syntax

    ```cs
    NewRelic.Api.Agent.NewRelic.SetApplicationName(string $name[, string $name_2, string $name_3])
    ```

    Set the app name for data rollup.

    ### Requirements

>     * Agent version 5.0.136.0 or higher.

Reviewer: 5.0.136.0 is EOL ( https://deploy-preview-17266--docs-website-netlify.netlify.app/docs/apm/agents/net-agent/getting-started/net-agent-eol-policy/ ), shouldn't this be all agent versions?

        </th>
        </tr>
    </thead>

    <tbody>
        <tr>
        <td>
            `$name`

            _string_
        </td>

        <td>
            Required. The primary application name.
        </td>
        </tr>

        <tr>
        <td>
            `$name_2`

            `$name_3`

            _string_
        </td>

        <td>
            Optional. Second and third names for app rollup. For more information, see [Use multiple names for an app](/docs/agents/manage-apm-agents/app-naming/use-multiple-names-app).
        </td>
        </tr>
    </tbody>
    </table>

    ### Examples

    ```cs
    NewRelic.Api.Agent.NewRelic.SetApplicationName("AppName1", "AppName2");
    ```
    </Collapser>
    <Collapser
        id="SetErrorGroupCallback"
        title="SetErrorGroupCallback"
    >
    ### Syntax

    ```cs
    NewRelic.Api.Agent.NewRelic.SetErrorGroupCallback(Func<IReadOnlyDictionary<string,object>, string> errorGroupCallback);
    ```
    Provide a callback method that takes an `IReadOnlyDictionary<string,object>` of attribute data, and returns an error group name.

    ### Requirements [#requirements]

    This API call is compatible with:

>     * Agent versions >= 10.9.0

Reviewer: Other calls say "or higher" rather than use puncutation

            `$account_value`

            _string_
        </td>

        <td>
            Required (can be null). Specify the name of a user account to associate with this page view. This value is assigned to the `account` key.
        </td>
        </tr>

        <tr>
        <td>
            `$product_value`

            _string_
        </td>

        <td>
            Required (can be null). Specify the name of a product to associate with this page view. This value is assigned to the `product` key.
        </td>
        </tr>
    </tbody>
    </table>

    ### Examples

    #### Record three user attributes [#report-three]

    ```cs
    NewRelic.Api.Agent.NewRelic.SetUserParameters("MyUserName", "MyAccountName", "MyProductName");
    ```

    #### Record two user attributes and one empty attribute [#report-two]

    ```cs
    NewRelic.Api.Agent.NewRelic.SetUserParameters("MyUserName", "", "MyProductName");
    ```

    </Collapser>
    <Collapser
        id="StartAgent"
        title="StartAgent"
    >
    ### Syntax

    ```cs
    NewRelic.Api.Agent.NewRelic.StartAgent()
    ```

    Start the agent if it hasn't already started. Usually unnecessary.

    ### Requirements

>     * Agent version 5.0.136.0 or higher.

Reviewer: Same version comment as previous, this is EOl version

    <Collapser
        id="GetLinkingMetadata"
        title="GetLinkingMetadata"
    >
    ### Syntax

    ```cs
    NewRelic.Api.Agent.NewRelic.GetLinkingMetadata();
    ```

    Returns key/value pairs which can be used to link traces or entities.

    ### Requirements

    * Agent version 8.19 or higher.
    * Compatible with all app types.

    ### Description

    The dictionary of key/value pairs returned includes items used to link traces and entities in the APM product. It will only contain items with meaningful values. For instance, if distributed tracing is disabled, `trace.id` will not be included.

    ### Return values

    `Dictionary <string, string>()` returned includes items used to link traces and entities in the APM product.

    ### Examples

    ```cs
    NewRelic.Api.Agent.IAgent Agent = NewRelic.Api.Agent.NewRelic.GetAgent();
    var linkingMetadata = Agent.GetLinkingMetadata();
    foreach (KeyValuePair<string, string> kvp in linkingMetadata)
    {
        Console.WriteLine("Key = {0}, Value = {1}", kvp.Key, kvp.Value);
    }
    ```

    </Collapser>
    <Collapser
        id="IAgent"
        title="IAgent"
    >
        ### Syntax

            ```cs

>             public interfac IAgent

Reviewer 2: Typo? 
```suggestion
            public interface IAgent
```
```suggestion
            public interfac IAgent
```

                Description
            </th>
            </tr>
        </thead>

        <tbody>
            <tr>
            <td>
                CurrentTransaction
            </td>

            <td>
                Property providing access to the currently executing transaction via the [ITransaction](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#ITransaction) interface. Must be called inside a [transaction](/docs/accounts-partnerships/education/getting-started-new-relic/glossary#transaction).
            </td>
            </tr>

            <tr>
            <td>
                CurrentSpan
            </td>

            <td>
                Property providing access to the currently executing span via the [ISpan](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#ISpan) interface.
            </td>
            </tr>
        </tbody>
        </table>

        ### Examples

        ```cs
        IAgent agent = NewRelic.Api.Agent.NewRelic.GetAgent();
        ITransaction transaction = agent.CurrentTransaction;
        ```

        </Collapser>
        <Collapser
            id="ITransaction"
            title="ITransaction"
        >
    ### Syntax

    ```cs

>     public interfac ITransaction

Reviewer 2: Typo? 

```suggestion
    public interface ITransaction
```
```suggestion
    public interfac ITransaction
```

        </td>

        <td>
            Add contextual information from your application to the current transaction in form of attributes.
        </td>
        </tr>

        <tr>
        <td>
            [`CurrentSpan`](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#CurrentSpan)
        </td>

        <td>
            Provides access to the currently executing [span](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#ISpan), which provides access to span-specific methods in the New Relic API.
        </td>
        </tr>

        <tr>
        <td>
            [`SetUserId`](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#SetUserId)
        </td>

        <td>
            Associates a user ID to the current transaction.
        </td>
        </tr>
    </tbody>
    </table>
    <CollapserGroup>
    <Collapser
        id="AcceptDistributedTraceHeaders"
        title="AcceptDistributedTraceHeaders"
    >
        ### Syntax

        ```cs
        void AcceptDistributedTraceHeaders(carrier, getter, transportType)
        ```

        Instrument the called service for inclusion in a distributed trace.

        ### Description

>         `ITransaction.AcceptDistributedTraceHeaders` is used to links the spans in a trace by accepting a payload generated by [`InsertDistributedTraceHeaders`](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#InsertDistributedTraceHeaders) or generated by some other W3C Trace Context compliant tracer. This method accepts the headers of an incoming request, looks for W3C Trace Context headers, and if not found, falls back to New Relic distributed trace headers. This method replaces the deprecated [`AcceptDistributedTracePayload`](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#InsertDistributedTraceHeaders) method, which only handles New Relic distributed trace payloads.

Reviewer 2: ```suggestion
        `ITransaction.AcceptDistributedTraceHeaders` is used to links the spans in a trace by accepting a payload generated by [`InsertDistributedTraceHeaders`](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#InsertDistributedTraceHeaders) or generated by some other W3C Trace Context compliant tracer. This method accepts the headers of an incoming request, looks for W3C Trace Context headers, and if not found, falls back to New Relic distributed trace headers. This method replaces the deprecated [`AcceptDistributedTracePayload`](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#AcceptDistributedTracePayload) method, which only handles New Relic distributed trace payloads.
```
Maybe add a sentence about the new/recommended one after this

    </Collapser>
    <Collapser
        id="SetUserId"
        title="SetUserId"
    >

    ### Syntax

    ```cs
    ITransaction SetUserId(string userId)
    ```

    Associates a user ID with the current transaction.

    This method requires .NET agent and .NET agent API [version 10.9.0](/docs/release-notes/agent-release-notes/net-release-notes/net-agent-1090) or higher.

    ### Parameters

    <table>
    <thead>
        <tr>
        <th>
            Parameter
        </th>

        <th>
            Description
        </th>
        </tr>
    </thead>

    <tbody>
        <tr>
        <td>
            `userId`

            _string_
        </td>

        <td>
            The User Id to be associated with this transaction.

            * `null`, empty and whitespace values will be ignored.
        </td>
        </tr>

    </tbody>
    </table>

    ### Example

    ```cs
    IAgent agent = NewRelic.Api.Agent.NewRelic.GetAgent(); 
    ITransaction transaction = agent.CurrentTransaction; 
    transaction.SetUserId("BobSmith123");
    ```
    </Collapser>
</CollapserGroup>
    </Collapser>
    <Collapser
        id="ISpan"
        title="ISpan"
    >
    ### Syntax

    ```cs

>     Public interfac ISpan

Reviewer 2: Typo? 

```suggestion
    Public interfac ISpan
```

    ```

    Provides access to span-specific methods in the New Relic API.

    ### Description

    Provides access to span-specific methods in the New Relic .NET agent API. To obtain a reference to `ISpan`, use:

    * The `CurrentSpan` property on [`IAgent`](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#IAgent) (Recommended).
    * The `CurrentSpan` property on [`ITransaction`](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#ITransaction).

    This section contains descriptions and parameters of `ISpan` methods:

    <table>
    <thead>
        <tr>
        <th>
            Name
        </th>

        <th>
            Description
        </th>
        </tr>
    </thead>

    <tbody>
        <tr>
        <td>

>             [`AddCustomAttribute`](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#add-custom-attr)

Reviewer 2: ```suggestion
            [`AddCustomAttribute`](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#AddCustomAttribute)
```

        </td>

        <td>
            Add contextual information from your application to the current span in form of attributes.
        </td>
        </tr>
        <tr>
        <td>
            [`SetName`](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#setname)
        </td>

        <td>
            Changes the name of the current span/segment/metrics that will be reported to New Relic.
        </td>
        </tr>
    </tbody>
    </table>

<CollapserGroup>
    <Collapser

>         id="add-custom-attr"

Reviewer 2: ```suggestion
        id="AddCustomAttribute"
```

    </tbody>
    </table>

    ### Examples

    ```cs
    NewRelic.Api.Agent.NewRelic.IncrementCounter("Custom/ExampleMetric");
    ```
    </Collapser>

    <Collapser
        id="NoticeError"
        title="NoticeError"
    >
    ### Overloads [#overloads]

    Notice an error and report to New Relic, along with optional custom attributes.

    ```cs
    NewRelic.Api.Agent.NewRelic.NoticeError(Exception $exception);
    NewRelic.Api.Agent.NewRelic.NoticeError(Exception $exception, IDictionary<TKey, TValue> $attributes);
    NewRelic.Api.Agent.NewRelic.NoticeError(string $error_message, IDictionary<TKey, TValue> $attributes);
    NewRelic.Api.Agent.NewRelic.NoticeError(string $error_message, IDictionary<TKey, TValue> $attributes, bool $is_expected);
    ```

    ### Requirements 

    This API call is compatible with:

    * All agent versions
    * All app types

    ### Description 

    Notice an error and report it to New Relic along with optional custom attributes. For each transaction, the agent only retains the exception and attributes from the first call to `NoticeError()`. You can pass an actual exception, or pass a string to capture an arbitrary error message.

>     If this method is invoked within a [transaction](/docs/glossary/glossary/#transaction), the agent reports the exception within the parent transaction. If it is invoked outside of a transaction, the agent creates an [error trace](/docs/errors-inbox/errors-inbox) and categorizes the error in the New Relic UI as a `NewRelic.Api.Agent.NoticeError` API call. If invoked outside of a transaction, the `NoticeError()` call will not contribute to the error rate of an application.

Reviewer 2: There's nothing on the errors inbox page about error trace

>     The agent adds the attributes only to the traced error; it does not send them to New Relic. For more information, see [`AddCustomAttribute()`](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#add-custom-attr).

Reviewer 2: ```suggestion
    The agent adds the attributes only to the traced error; it does not send them to New Relic. For more information, see [`AddCustomAttribute()`](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#AddCustomAttribute).
```

            <td>
                Required. The `Exception` you want to instrument. Only the first 10,000 characters from the stack trace are retained.
            </td>
            </tr>

            <tr>
            <td>
                `$attributes`

                _IDictionary&lt;TKey, TValue&gt;_
            </td>

            <td>
                Specify key/value pairs of attributes to annotate the error message. The `TKey` must be a string, the `TValue` can be a string or object.
            </td>
            </tr>
        </tbody>
        </table>
    </Collapser>
    <Collapser
        id="NoticeError(String, IDictionary)"
        title="string-idictionary-overload"
    >
        ```cs
        NewRelic.Api.Agent.NewRelic.NoticeError(string $error_message, IDictionary<TKey, TValue> $attributes)
        ```

        <table>
        <thead>
            <tr>
            <th width="25%">
                Parameter
            </th>

            <th>
                Description
            </th>
            </tr>
        </thead>

        <tbody>
            <tr>
            <td>
                `$error_message`

                _string_
            </td>

            <td>

>                 Required. Specify a string to report to New Relic as though it's an exception. This method creates both [error events and error traces](/docs/tutorial-error-tracking/respond-outages/).  Only the first 1023 characters are retained in error events, while error traces retain the full message.

Reviewer 2: There's nothing about error events/error traces on that doc

    };

    NewRelic.Api.Agent.NewRelic.SetErrorGroupCallback(errorGroupCallback);
    ```

    </Collapser>
    <Collapser
        id="SetTransactionName"
        title="SetTransactionName"
    >
    ### Syntax

    ```cs
    NewRelic.Api.Agent.NewRelic.SetTransactionName(string $category, string $name)
    ```

    Sets the name of the current transaction.

    ### Requirements

    * Compatible with all agent versions.
    * Must be called inside a [transaction](/docs/accounts-partnerships/education/getting-started-new-relic/glossary#transaction).

    ### Description

    Sets the name of the current transaction. Before you use this call, ensure you understand the implications of [metric grouping issues](/docs/agents/manage-apm-agents/troubleshooting/metric-grouping-issues).

    If you use this call multiple times within the same transaction, each call overwrites the previous call and the last call sets the name.

    <Callout variant="important">
    Do not use brackets `[suffix]` at the end of your transaction name. New Relic automatically strips brackets from the name. Instead, use parentheses `(suffix)` or other symbols if needed.
    </Callout>

>     Unique values like URLs, page titles, hex values, session IDs, and uniquely identifiable values should not be used in naming your transactions. Instead, add that data to the transaction as a custom parameter with the [`AddCustomAttribute()`](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#add-custom-attr) call.

Reviewer 2: ```suggestion
    Unique values like URLs, page titles, hex values, session IDs, and uniquely identifiable values should not be used in naming your transactions. Instead, add that data to the transaction as a custom parameter with the [`AddCustomAttribute()`](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#AddCustomAttribute) call.
```
```suggestion
    Unique values like URLs, page titles, hex values, session IDs, and uniquely identifiable values should not be used in naming your transactions. Instead, add that data to the transaction as a custom parameter with the [`AddCustomAttribute()`](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#add-custom-attr) call.
```

    <tbody>
        <tr>
        <td>
            `$category`

            _string_
        </td>

        <td>
            Required. The category of this transaction, which you can use to distinguish different types of transactions. Defaults to <DoNotTranslate>**`Custom`**</DoNotTranslate>. Only the first 255 characters are retained.
        </td>
        </tr>

        <tr>
        <td>
            `$name`

            _string_
        </td>

        <td>
            Required. The name of the transaction. Only the first 255 characters are retained.
        </td>
        </tr>
    </tbody>
    </table>

    ### Examples

    ```cs
    NewRelic.Api.Agent.NewRelic.SetTransactionName("Other", "MyTransaction");
    ```

    </Collapser>
    <Collapser
        id="SetTransactionUri"
        title="SetTransactionUri"
    >
    ### Syntax

    ```cs
    NewRelic.Api.Agent.NewRelic.SetTransactionUri(Uri $uri)
    ```

    Sets the URI of the current transaction.

    ### Requirements

    * Must be called inside a [transaction](/docs/accounts-partnerships/education/getting-started-new-relic/glossary#transaction).
    * Agent version 6.16 or higher.

    <Callout variant="important">

>     This method only works when used within a transaction created using the `Transaction` attribute with the `Web` property set to `true`. (See [Instrument using attributes](/docs/agents/net-agent/api-guides/net-agent-api-instrument-using-attributes).) It provides support for custom web-based frameworks that the agent does not automatically support.

Reviewer 2: ```suggestion
    This method only works when used within a transaction created using the `Transaction` attribute with the `Web` property set to `true` (see [Custom instrumentation via attributes](/docs/agents/net-agent/api-guides/net-agent-api-instrument-using-attributes)). It provides support for custom web-based frameworks that the agent does not automatically support.
```

    </Callout>

    ### Description

>     Sets the URI of the current transaction. The URI appears in the 'request.uri' [attribute](/docs/agents/net-agent/attributes/net-agent-attributes) of [transaction traces](/docs/apm/transactions/transaction-traces/transaction-traces) and [transaction events](/docs/using-new-relic/metrics/analyze-your-metrics/data-collection-metric-timeslice-event-data), and also can affect transaction naming.

Reviewer 2: ```suggestion
    Sets the URI of the current transaction. The URI appears in the `request.uri` attribute of [transaction traces](/docs/apm/transactions/transaction-traces/transaction-traces) and [transaction events](/docs/using-new-relic/metrics/analyze-your-metrics/data-collection-metric-timeslice-event-data), and it also can affect transaction naming.
```

    If you use this call multiple times within the same transaction, each call overwrites the previous call. The last call sets the URI.

    <DoNotTranslate>**Note**</DoNotTranslate>: as of agent version 8.18, the `request.uri` attribute's value is set to the value of the `Uri.AbsolutePath` property of the `System.Uri` object passed to the API.

    ### Parameters

    <table>
    <thead>
        <tr>
        <th width="25%">
            Parameter
        </th>

        <th>
            Description
        </th>
        </tr>
    </thead>

    <tbody>
        <tr>
        <td>
            `$uri`

            _Uri_
        </td>

        <td>
            The URI of this transaction.
        </td>
        </tr>
    </tbody>
    </table>

    ### Examples

    ```cs
    var uri = new System.Uri("https://www.mydomain.com/path");
    NewRelic.Api.Agent.NewRelic.SetTransactionUri(uri);
    ```

    </Collapser>
    <Collapser
        id="SetUserParameters"
        title="SetUserParameters"
    >
    ### Syntax

    ```cs
    NewRelic.Api.Agent.NewRelic.SetUserParameters(string $user_value, string $account_value, string $product_value)
    ```

>     Create user-related custom attributes. [`AddCustomAttribute`](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#add-custom-attr) is more flexible.

Reviewer 2: ```suggestion
    Create user-related custom attributes. [`AddCustomAttribute`](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#AddCustomAttribute) is more flexible.
```

    ### Requirements

    * Compatible with all agent versions.
    * Must be called inside a [transaction](/docs/accounts-partnerships/education/getting-started-new-relic/glossary#transaction).

    ### Description

    <Callout variant="tip">

>     This call only allows you to assign values to pre-existing keys. For a more flexible method to create key/value pairs, use [`AddCustomAttribute()`](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#add-custom-attr).

Reviewer 2: ```suggestion
    This call only allows you to assign values to pre-existing keys. For a more flexible method to create key/value pairs, use [`AddCustomAttribute()`](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#AddCustomAttribute).
```

    </Callout>

>     Define user-related [custom attributes](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#add-custom-attr)to associate with a browser page view (user name, account name, and product name). The values are automatically associated with pre-existing keys (`user`, `account`, and `product`), then attached to the parent APM transaction. You can also [attach (or "forward") these attributes](/docs/insights/new-relic-insights/decorating-events/insights-custom-attributes#forwarding-attributes) to browser [PageView](/docs/agents/manage-apm-agents/agent-metrics/agent-attributes#destinations) events.

Reviewer 2: ```suggestion
    Define user-related [custom attributes](/docs/apm/agents/net-agent/net-agent-api/net-agent-api/#AddCustomAttribute)to associate with a browser page view (user name, account name, and product name). The values are automatically associated with pre-existing keys (`user`, `account`, and `product`), then attached to the parent APM transaction. You can also [attach (or "forward") these attributes](/docs/insights/new-relic-insights/decorating-events/insights-custom-attributes#forwarding-attributes) to browser [PageView](/docs/agents/manage-apm-agents/agent-metrics/agent-attributes#destinations) events.
```
